kerrighed-2.4.4/man/krgadm.1
'\" t
.\"     Title: krgadm
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/29/2010
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "KRGADM" "1" "01/29/2010" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
krgadm \- Start, stop or get status of a Kerrighed cluster
.SH "SYNOPSIS"
.HP \w'\fBkrgadm\fR\ 'u
\fBkrgadm\fR [\-h | \-\-help]
.HP \w'\fBkrgadm\fR\ 'u
\fBkrgadm\fR cluster {status | wait_start | poweroff | reboot}
.HP \w'\fBkrgadm\fR\ 'u
\fBkrgadm\fR cluster {start} [\-n | \-\-nodes] [\-a | \-\-auto]
.HP \w'\fBkrgadm\fR\ 'u
\fBkrgadm\fR nodes
.SH "DESCRIPTION"
.PP

\fBkrgadm\fR
allows the user to start, stop or get status of a Kerrighed cluster\&.\&.
.PP
Each
\fBkrgadm\fR
invocation is followed by a command and some options relative to this mode\&.
.PP
cluster
.RS 4
Start, stop or see the cluster status\&.
.RE
.PP
nodes
.RS 4
Get cluster nodes status\&.
.RE
.SH "GENERAL OPTIONS"
.PP
.PP
\-h, \-\-help
.RS 4
Print help and exit
.RE
.SH "MODES"
.PP
.PP
cluster
.RS 4
Manage the cluster
.RE
.PP
nodes
.RS 4
Prints cluster nodes status
.RE
.SH "CLUSTER COMMANDS"
.PP
.PP
start
.RS 4
Start the cluster (see CLUSTER START OPTIONS below)
.RE
.PP
status
.RS 4
Print cluster status
.RE
.PP
wait_start
.RS 4
Return once cluster is started
.RE
.PP
poweroff
.RS 4
Power off all nodes belonging to the cluster
.RE
.PP
reboot
.RS 4
Reboot all nodes belonging to the cluster
.RE
.SH "CLUSTER START OPTIONS"
.PP
By default,
\fBkrgadm cluster start\fR
start the cluster with available nodes\&.
.PP
.PP
\-n, \-\-nodes
.RS 4
Wait for a given list of nodes before starting\&. List of nodes is of the form: 1,3,7,10\-12,43\-46
.RE
.PP
\-n, \-\-nodes
.RS 4
Wait for the given number of nodes to be present before starting\&.
.RE
.SH "AUTHOR"
.PP
Renaud Lottiaux
renaud\&.lottiaux@kerlabs\&.com
.PP
Jean Parpaillon
jean\&.parpaillon@kerlabs\&.com


kerrighed-2.4.4/ChangeLog
Kerrighed version 2.4.4
   Features:
   * Port driver for Realtek R8169 (present on some Atom platforms)
   * Update tg3 driver from Broadcom website. Add support for BCM5784 for instance
   * Upgrade krgadm
   Fixes:
   * Check for admin right for cluster start/shutdown/restart

Kerrighed version 2.4.2
   Fixes:
   * Fix crashes when doing CTRL-C on a process doing memory injection.
   * Fix some random memory incoherence in process distant fork.
   * Fix OOM in injection mechanism on large memory systems.
   * Avoid migration of pages from processes without USE_REMOTE_MEMORY.
   * Fix some crashes during distant fork or migration on loaded memory system.

Kerrighed version 2.4.1
   Fixes:
   * Major performance improvement
   * Fix crashes during migration or distant fork using Lustre.
   * Fix crashes with distributed system V memory segments.
   * Fix dead-lock in the MM sub-system
   * Fix issues in system V message queue identifiers
   * Fix deadlock on 32bit platforms in the RPC layer
   * Fix forgotten processes in /proc/<pid> directories
   * Fix OOM issue using injection mechanism
   * Fix memory allocation failure when big buffers are passed to read or write
   * Various memory leak fixes.
   * Various fixes in the checkpoint/restart sub-system
   * Various other fixes.

Kerrighed version 2.4.0
   Features:
   * Remote memory injection
   * Configuration scheduler
   * Application check-pointing
   * Remote vfork()
   * New RPC transmission engine that highly increases performance

   Fixes:
   * Fix random seg-fault during process migration on 64bits machines.
   * Fix incompatibilities with IPv6 tunnels and other special net devices
   * Various bugfixes in RPCs
   * Fix a strong memory leak
   * Fix wrong attempts to migrate threads when this is not supported
   * Fix random userspace memory access when using fcntl on files remotely open
   * Fix random userspace memory access when using ioctl on files remotely open
   * Fix BUG when the parent of an exiting multi-threaded process is/was remote
   * Fix many socket related bugs in the context of migrated processes
   * Fix many bugs in krgadm (seg-fault, etc)

   Build system:
   * Remove sub-configures
   * Split tools dir into scripts/ man/ and tools/
   * Create patched kernel source dir in kernel/, no need for --with-kernel option
   * Check architecture compatibility

Kerrighed version 2.3.0
   Fixes:
   * Fix uninitialized anonymous memory mapping
   * Limit network bandwidth consumption due to abusive packet retransmissions
   * Fix root user's ability to kill other's processes when they are remote
   * Fix bug in FAF version of recv and recvfrom
   * Fix user-level deadlocks when interrupting read/write operations with FAF
   * Fix build issues on systems having not sh pointing to bash
   * Fix compile error with debugs and debugfs disabled
   * Fix BUG when processes having multi-threaded children have remote children
   * Fix crash when the scheduler attempts to migrate dead tasks
   * Fix compile error on x86_64
   * Fix unpredictable locking behavior due to the use of a wrong prototype
   * Fix memory leak of slab obj_entry.
   * Fix rare close_faf_file BUG during process migration / distant fork.
   * Fix various signal issues in the FAF mechanism.
   * Fix various signal issues in the communication layer.
   * Fix permission checking when sending a signal to a remote task
   * Fix compilation when building kernel with CONFIG_HIGHMEM_64 on x86_32
   * Add support for disabled CONFIG_COMPAT_VDSO option.
   * Fix system V SHM in SMP mode
   * Fix system V FIFO in SMP mode
   * Fix system V SEM in SMP mode
   * Fix craches related to the use of the recvmsg function

Kerrighed version 2.2.1
   Fixes:
   * Fix crash in fcntl64 with recent libc
   * Fix memory leak in migration and remote fork
   * Fix access to freed memory after migration
   * Fix NUL byte inserted in the first line of /proc/stat
   * Fix memory leaks in FAFed recvmsg() and exit of remote forked processes
   * Fix deadlock when killing non existing (but realistic) pgrp

Kerrighed version 2.2.0
   * SMP support
   * Full distributed IPC support
   * Lots of bug fixes.

Kerrighed version 2.1.0

   * Port to 2.6.20
   * Rewritting of the communication layer on top of TIPC.
   * Lots of bug fixes.

Kerrighed version 2.0.0

   Strong refactoring and janitoring.
   This version is the basis for our future developments.
   Less features than in the previous versions, but more stability and
   much cleaner code.
   Features removed in this version will be re-enabled in the next
   releases.

   Features removed temporarily:
   * KerFS
   * Distributed threads
   * Checkpoint
   * Dynamic streams.

Kerrighed version 1.1.0-beta

   Features:
   * Port to Linux 2.6.11.

Kerrighed version 1.0.2
   Fixes:
   * Fix random signal delivering in some cases.
   * Fix useless network traffic when some applications are running
     (xosview for instance).
   * Fix some deadlock cases in the global lock mechanism.
   * Fix a kernel ooops in distributed thread brk function.
   * Fix large memory leak with Kerrighed threads and process migration.
   * Fix some distributed lock problems using Kerrighed threads.
   * Fix a memory leak in KerFS.

   Features:
   * Enable to see on which node a process is running in the top/ps commands.

Kerrighed version 1.0.1
   Fixes:
   * Remove deprecated dependancy on PERFCTNR.
   * Fix crash with global INET socket in some corner cases.
   * Fix major issue when HIGHMEM support is activated.

   Features:
   * Port to Linux 2.4.29.
   * Enable to use Kerrighed threads through LD_PRE_LOAD.

Kerrighed version 1.0.0
   Fixes:
   * Add permission check on commands "migrate" and "checkpoint".
   * Avoid process migration on a non existing node !
   * Fix up cluster wide PIDs problems for threads.
   * Fix crash during process migration with containers in some corner cases.

kerrighed-2.4.4/tools/krgadm.c
 /*
 *  Copyright (C) 2006-2007, Pascal Gallard, Kerlabs.
 *  Copyright (c) 2008 Jean Parpaillon <jean.parpaillon@kerlabs.com>
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>

#include <kerrighed.h>

enum {
	NONE,
	STATUS,
	START,
	WAIT_START,
	POWEROFF,
	REBOOT,
};

#define NODE_SEP ','
#define NODE_RANGE_SEP '-'
#define POLL_NODES 1
#define NODES_OPTION "nodes", required_argument, NULL, 'n'
#define AUTO_OPTION "auto", required_argument,NULL, 'a'

static struct option start_options[] = {
  {NODES_OPTION},
  {AUTO_OPTION},
  {NULL, 0, NULL, 0}
};

void help(char * program_name)
{
	printf("\
Usage: %s -h,--help\n\
  or:  %s cluster {status|wait_start|poweroff|reboot}\n\
  or:  %s cluster start  [-n|--nodes] [-a|--auto]\n\
  or:  %s nodes\n",
	       program_name, program_name, program_name, program_name);
	printf("\n\
Mode:\n\
  cluster           clusters management\n\
  nodes             print nodes status\n\
\n\
Cluster start options:\n\
  -n, --nodes       list of nodes to wait for before starting\n\
                    ie: 2,3,6-10,42-45\n\
  -a, --auto        number of nodes to wait for before starting\n\
\n\
Node Status:\n\
  present           available for integrating the cluster\n\
  online            participating in the cluster\n\
\n");
}

int parse_node(char *ch, struct krg_node_set *node_set)
{
	int node;
	char* endptr;
	int r = 1;

	node = strtol(ch, &endptr, 10);
	if (endptr == ch || *endptr) {
		r = -1;
		errno = EINVAL;
		perror("node");
	} else
		krg_node_set_add(node_set, node);

	return r;
}

int parse_nodes_interval(char *ch, struct krg_node_set *node_set)
{
	int node1, node2, node;
	char* index;
	char* endptr;
	int r = 1;

	index = strchr(ch, NODE_RANGE_SEP);
	node1 = strtol(ch, &endptr, 10);
	if (endptr == ch || NODE_RANGE_SEP != *endptr) {
		r = -1;
		errno = EINVAL;
		perror("first node range");
		goto exit;
	}
	node2 = strtol(index+1, &endptr, 10);
	if (endptr == index+1 || *endptr) {
		r = -1;
		errno = EINVAL;
		perror("end node range");
		goto exit;
	}

	for(node=node1; node<=node2; node++)
		krg_node_set_add(node_set, node);

 exit:
	return r;
}

int parse_nodes(char *ch, struct krg_node_set *node_set)
{
	char *ch_item;
	int r = 1;

	ch_item = strrchr(ch, NODE_SEP);

	if(ch_item){
		*ch_item = 0;
		if ((r=parse_nodes(ch, node_set)) == -1)
			goto exit;
		if ((r=parse_nodes(ch_item+1, node_set)) == -1)
			goto exit;
		goto exit;
	}

	ch_item = strrchr(ch, NODE_RANGE_SEP);
	if(ch_item) {
		if ((r=parse_nodes_interval(ch, node_set)) == -1)
			goto exit;
	} else
		if ((r=parse_node(ch, node_set)) == -1)
			goto exit;

 exit:
	return r;
}

/*
 * Return number of nodes in the cluster if up, 0 if not, -1 on failure
 */
int cluster_status()
{
	struct krg_clusters* cluster_status;
	int i = 0;

	cluster_status = krg_cluster_status();
	if(!cluster_status){
		i = -1;
		goto exit;
	}

	if (krg_clusters_is_up(cluster_status, 0)){
		i = krg_nodes_num_online(krg_nodes_status());
	}

 exit:
	krg_clusters_destroy(cluster_status);
	return i;
}

/*
 * When returning on success, node_set contains nodes to start.
 *
 * Return 1 when i nodes are present, -1 on failure.
 */
int wait_for_nodes_count(int i, struct krg_node_set* node_set)
{
	struct krg_nodes* status;
	int cur, r = 1;
	int nodes_count;

	printf("Waiting for %d nodes to join... ", i);
	fflush(stdout);

	if (cluster_status() > 0) {
		r = -1;
		errno = EALREADY;
		goto exit;
	}

	if (i < 1 || i > krg_get_max_nodes()) {
		r = -1;
		errno = ERANGE;
		goto exit;
	}

	do {
		status = krg_nodes_status();
		if (!status) {
			r = -1;
			goto exit;
		}

		nodes_count = krg_nodes_num_present(status);
		printf("%4d/%-4d", nodes_count, i);
		fflush(stdout);

		if (nodes_count < i) {
			sleep(POLL_NODES);
			printf("\b\b\b\b\b\b\b\b\b");
			krg_nodes_destroy(status);
		}

		fflush(stdout);
	} while (nodes_count < i);

	nodes_count = 0;
	cur = -1;
	do {
		cur = krg_nodes_next_present(status, cur);
		krg_node_set_add(node_set, cur);
		nodes_count++;
	} while (nodes_count < i);

	krg_nodes_destroy(status);

 exit:
	if (r==1)
		printf(" done\n");
	else
		printf(" fail (%s)\n", strerror(errno));

	return r;
}

/*
 * Return 1 when nodes in node_set are present, -1 on failure.
 */
int wait_for_nodes(struct krg_node_set* node_set)
{
	struct krg_nodes* status;
	int bcl, r = 1, done, node_count;

	printf("Waiting for nodes to join... ");
	fflush(stdout);

	if (cluster_status() > 0) {
		r = -1;
		errno = EALREADY;
		goto exit;
	}

	node_count = krg_node_set_weight(node_set);
	do {
		done = 1;
		status = krg_nodes_status();
		if (!status) {
			r = -1;
			goto exit;
		}
		for (bcl = 0; bcl < krg_get_max_nodes(); bcl++) {
			if (krg_node_set_contains(node_set, bcl)) {
				printf("%4d:", bcl);
				if (krg_nodes_is_present(status, bcl))
					printf("1");
				else {
					done = 0;
					printf("0");
				}
			}
			fflush(stdout);
		}
		krg_nodes_destroy(status);

		if (!done) {
			sleep(POLL_NODES);

			for (bcl = 0; bcl < node_count; bcl++)
				printf("\b\b\b\b\b\b");
			fflush(stdout);
		}

	} while (!done);

 exit:
	if (r==1)
		printf(" done\n");
	else
		printf(" fail (%s)\n", strerror(errno));

	return r;
}

/*
 * Return 0 on success, -1 on failure
 */
int cluster_start(int argc, char* argv[], char* program_name)
{
	struct krg_node_set* node_set;
	int c, option_index;
	int r = 0;
	int wait = -1;
	char* nodes;
	char* endptr;

	node_set = krg_node_set_init();

	while ((c=getopt_long(argc, argv, "n:a:",
												start_options, &option_index)) != -1) {
		switch (c) {
		case 'n':
			wait = 0;
			nodes = malloc(sizeof(char)*(strlen(optarg)+1));
			nodes = strncpy(nodes, optarg, strlen(optarg)+1);
			if ((r=parse_nodes(optarg, node_set)) == -1)
				goto exit;
			break;
		case 'a':
			wait = strtol(optarg, &endptr, 10);
			if (endptr == optarg || *endptr) {
				r = -1;
				errno = EINVAL;
				perror("nodes number");
				goto exit;
			}
			break;
		default:
			help(program_name);
			goto exit;
		}
	}

	switch (wait) {
	case -1:
		{
			struct krg_nodes* status;

			status = krg_nodes_status();
			node_set = krg_nodes_get_present(status);
			krg_nodes_destroy(status);
		}
		break;
	case 0:
		r = wait_for_nodes(node_set);
		if (r == -1)
			goto exit;
		break;
	default:
		r = wait_for_nodes_count(wait, node_set);
		if (r == -1) {
			goto exit;
		}
	}

	printf("Starting cluster... ");
	fflush(stdout);
	r = krg_cluster_start(node_set);
	if (r == -1)
		printf("fail (%s)\n", strerror(errno));
	else
		printf("done\n");

 exit:
	return r;
}

/*
 * Return EXIT_SUCCESS on success, EXIT_FAILURE on failure.
 */
int cluster(int argc, char* argv[], char* program_name)
{
	int action = NONE;
	int ret = EXIT_SUCCESS;
	int r;

	if(argc==0 || !strcmp(*argv, "status"))
		action = STATUS;
	else if(!strcmp(*argv, "start"))
		action = START;
	else if(!strcmp(*argv, "wait_start"))
		action = WAIT_START;
	else if(!strcmp(*argv, "poweroff"))
		action = POWEROFF;
	else if(!strcmp(*argv, "reboot"))
		action = REBOOT;

	switch (action) {
	case STATUS:
		printf("status: ");
		r = cluster_status();
			switch(r){
			case -1:
				printf("error\n");
				ret = EXIT_FAILURE;
				break;
			case 0:
				printf("down\n");
				break;
			default:
				printf("up on %d nodes\n", r);
			}
			break;
	case START:
		r = cluster_start(argc, argv, program_name);
		if (r == -1)
			ret = EXIT_FAILURE;
		break;
	case WAIT_START:
		printf("Waiting for cluster to start... ");
		fflush(stdout);
		r = krg_cluster_wait_for_start();
		if (r == -1) {
			printf("fail (%s)\n", strerror(errno));
			ret = EXIT_FAILURE;
		} else
			printf("done\n");
		break;
	case POWEROFF:
		printf("Shutting down cluster... ");
		fflush(stdout);
		if (cluster_status()==0) {
			printf("fail (cluster not running)\n");
			ret = EXIT_FAILURE;
		} else {
			r = krg_cluster_shutdown(0);
			if (r == -1) {
				printf("fail (%s)\n", strerror(errno));
				ret = EXIT_FAILURE;
			} else
				printf("done\n");
		}
		break;
	case REBOOT:
		printf("Rebooting cluster... ");
		fflush(stdout);
		if (cluster_status() == 0) {
			printf("fail (cluster not running)\n");
			ret = EXIT_FAILURE;
		} else {
			r = krg_cluster_reboot(0);
			if (r == -1) {
				printf("fail (%s)\n", strerror(errno));
				ret = EXIT_FAILURE;
			} else
				printf("done\n");
		}
	break;
	default:
		help(program_name);
	}

	return ret;
}

/*
 * Return EXIT_SUCCESS on success, EXIT_FAILURE on error
 */
int nodes(int argc, char* argv[], char* program_name)
{
	int ret = EXIT_SUCCESS;

	if(argc==0 || !strcmp(*argv, "status")){
		struct krg_nodes* status;
		int bcl;

		status = krg_nodes_status();

		if(status){
			for(bcl=0;bcl<krg_get_max_nodes();bcl++)
				if(status->nodes[bcl]!=INVALID)
					printf("%d:%s\n", bcl, krg_status_str(status->nodes[bcl]));
		}else{
			printf("Are you sure to run Kerrighed ?\n");
			ret = EXIT_FAILURE;
		}
	}else
		help(program_name);

	return ret;
}

int main(int argc, char* argv[])
{
	char **arg;
	char* program_name;
	int count;
	int ret = EXIT_SUCCESS;

	arg = argv;
	count = argc;
	program_name = argv[0];

	if(argc==1){
		help(program_name);
		exit(EXIT_SUCCESS);
	}

	opterr = 0;

	count--;
	arg++;

	if (!strcmp(argv[1], "cluster"))
		ret = cluster(count-1, arg+1, program_name);
	else if (!strcmp(argv[1], "nodes"))
		ret = nodes(count-1, arg+1, program_name);
	else
		help(program_name);

	return ret;
}
